// @blockone/eosjs-ledger-signature-provider/dist/SignatureProvider.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var LedgerAPI_1 = require("./LedgerAPI");
var LedgerUtils_1 = require("./LedgerUtils");
var SignatureProvider = /** @class */ (function () {
    function SignatureProvider() {
        this.eosjsApi = null;
        this.ledgerApi = null;
        this.cachedKeys = [];
    }
    SignatureProvider.prototype.getLedgerApi = function () {
        return __awaiter(this, void 0, void 0, function () {
            var transport;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.ledgerApi) {
                            return [2 /*return*/, this.ledgerApi];
                        }
                        return [4 /*yield*/, LedgerUtils_1.getTransport()];
                    case 1:
                        transport = _a.sent();
                        this.ledgerApi = new LedgerAPI_1.LedgerAPI(transport);
                        return [2 /*return*/, this.ledgerApi];
                }
            });
        });
    };
    /** Public keys associated with the private keys that the `SignatureProvider` holds */
    SignatureProvider.prototype.getAvailableKeys = function (requestPermission) {
        return __awaiter(this, void 0, void 0, function () {
            var api, key, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.cachedKeys.length) {
                            return [2 /*return*/, this.cachedKeys];
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 4, , 5]);
                        return [4 /*yield*/, this.getLedgerApi()];
                    case 2:
                        api = _a.sent();
                        return [4 /*yield*/, api.getPublicKey(requestPermission)];
                    case 3:
                        key = _a.sent();
                        this.cachedKeys = [key];
                        return [2 /*return*/, this.cachedKeys];
                    case 4:
                        error_1 = _a.sent();
                        throw error_1;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /** Sign a transaction */
    SignatureProvider.prototype.sign = function (_a) {
        var chainId = _a.chainId, serializedTransaction = _a.serializedTransaction;
        return __awaiter(this, void 0, void 0, function () {
            var api, signatures, error_2;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.getLedgerApi()];
                    case 1:
                        api = _b.sent();
                        return [4 /*yield*/, api.signTransaction({ chainId: chainId, serializedTransaction: serializedTransaction })];
                    case 2:
                        signatures = _b.sent();
                        return [2 /*return*/, { signatures: signatures, serializedTransaction: serializedTransaction }];
                    case 3:
                        error_2 = _b.sent();
                        throw error_2;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    SignatureProvider.prototype.getCachedKeys = function () {
        return this.cachedKeys;
    };
    SignatureProvider.prototype.setCachedKeys = function (keys) {
        this.cachedKeys = keys;
    };
    SignatureProvider.prototype.clearCachedKeys = function () {
        this.cachedKeys = null;
    };
    SignatureProvider.prototype.cleanUp = function () {
        return;
    };
    return SignatureProvider;
}());
exports.SignatureProvider = SignatureProvider;
