// @blockone/ual-ledger/src/Ledger.ts
import {
  Authenticator,
  ButtonStyle,
  Chain,
  UALAccountName,
  UALError,
  UALErrorType,
  UALLoggedInAuthType,
  User,
} from '@blockone/universal-authenticator-library'
import { CONSTANTS } from './constants'
import { LedgerOptions, Name } from './interfaces'
import { ledgerLogo } from './ledgerLogo'
import { LedgerUser } from './LedgerUser'
import { UALLedgerError } from './UALLedgerError'

export class Ledger extends Authenticator {
  private onBoardingLink: string = CONSTANTS.onBoardingLink
  private users: LedgerUser[] = []
  public chains: Chain[]
  public options?: LedgerOptions

  constructor(chains: Chain[], options?: LedgerOptions) {
    super(chains, options)
    this.chains = chains
  }

  public async init(): Promise<void> {
    console.info('Ledger initialized!')
  }

  public shouldRender(): boolean {
    if (window.location.protocol !== 'https:') {
      return false
    }

    return true
  }

  public shouldAutoLogin(): boolean {
    return false
  }

  public async shouldRequestAccountName(): Promise<boolean> {
    return true
  }

  public async login(accountName?: string): Promise<User[]> {
    for (const chain of this.chains) {
      const user = new LedgerUser(chain, accountName, this.requiresGetKeyConfirmation(accountName))
      await user.init()
      const isValid = await user.isAccountValid()
      if (!isValid) {
        const message = `Error logging into account "${accountName}"`
        const type = UALErrorType.Login
        const cause = null
        throw new UALLedgerError(message, type, cause)
      }
      this.users.push(user)
    }

    return this.users
  }

  public async logout(): Promise<void> {
    try {
      for (const user of this.users) {
        user.signatureProvider.cleanUp()
        user.signatureProvider.clearCachedKeys()
      }
      this.users = []
    } catch (e) {
      const message = CONSTANTS.logoutMessage
      const type = UALErrorType.Logout
      const cause = e
      throw new UALLedgerError(message, type, cause)
    }
  }

  public getStyle(): ButtonStyle {
    return {
      icon: ledgerLogo,
      text: Name,
      textColor: CONSTANTS.white,
      background: CONSTANTS.ledgerGreen,
    }
  }

  public isLoading(): boolean {
    return false
  }

  public isErrored(): boolean {
    return false
  }

  public getError(): UALError | null {
    return null
  }

  public getOnboardingLink(): string {
    return this.onBoardingLink
  }

  public reset(): void {
    return
  }

  public requiresGetKeyConfirmation(accountName?: string): boolean {
    if (!accountName) {
      return true
    }

    const type = window.localStorage.getItem(UALLoggedInAuthType)
    const account = window.localStorage.getItem(UALAccountName)
    if (account === accountName && type === Name) {
      return false
    }

    return true
  }
}
